### PR TITLE
feat(metadata-sidebar): add FilterInstancesDropdown

### PR DESCRIPTION
**Pull request by kajarosz**
Implement `FilterInstancesDropdown` as subheader in Metadata Sidebar. Dropdown should be visible if user added more than one template instance. It allows user to filter template instances visible in view mode list.

Changes in `src/elements/content-sidebar/hooks/useSidebarMetadataFetcher.ts` are not part of this PR - this is copied from https://github.com/box/box-ui-elements/pull/3759 in order for fetcher to work properly. https://github.com/box/box-ui-elements/pull/3759 must be merged first.

### Screenshots

Screenshot 2024-12-02 at 13 00 49

Screenshot 2024-12-02 at 13 19 23

### CODE REVIEW

**Comment by wpiesiak** [comment]
Can you attach some visuals?

**Review by wpiesiak — APPROVED** [review]

File: src/elements/content-sidebar/MetadataSidebarRedesign.tsx [hunk 1]
```diff
@@ -156,11 +157,25 @@ function MetadataSidebarRedesign({ api, elementId, fileId, onError, isFeatureEna
...
+    const [filteredTemplates, setFilteredTemplates] = React.useState([]);
+    const filterDropdown = status === STATUS.SUCCESS && appliedTemplates.length > 1 && (
+        <FilterInstancesDropdown
+            appliedTemplates={appliedTemplates as MetadataTemplate[]}
+            selectedTemplates={filteredTemplates}
+            setSelectedTemplates={setFilteredTemplates}
+        />
+    );
+
+    const filteredTemplateIstances = templateInstances.filter(instance =>
```
Comment (JChan106) [inline src/elements/content-sidebar/MetadataSidebarRedesign.tsx:174]: typo: filteredTemplateInstances

File: src/elements/content-sidebar/MetadataSidebarRedesign.tsx [hunk 2]
```diff
@@ -156,11 +157,25 @@ function MetadataSidebarRedesign({ api, elementId, fileId, onError, isFeatureEna
...
         <AddMetadataTemplateDropdown
             availableTemplates={templates}
-            selectedTemplates={selectedTemplates as MetadataTemplate[]}
+            selectedTemplates={appliedTemplates as MetadataTemplate[]}
             onSelect={handleTemplateSelect}
         />
     );
 
+    const [filteredTemplates, setFilteredTemplates] = React.useState([]);
+    const filterDropdown = status === STATUS.SUCCESS && appliedTemplates.length > 1 && (
```
Comment (JChan106) [inline src/elements/content-sidebar/MetadataSidebarRedesign.tsx:166]: Lets make sure this passes null if the conditional fails

File: src/elements/content-sidebar/SidebarContent.js [hunk 3]
```diff
@@ -33,6 +34,7 @@ const SidebarContent = ({ actions, children, className, elementId, sidebarView,
                 {title && <h3 className="bcs-title">{title}</h3>}
                 {actions}
             </div>
+            {subheader && <div className="bcs-content-subheader">{subheader}</div>}
```
Comment (JChan106) [inline src/elements/content-sidebar/SidebarContent.js:37]: I see we pass in children below this. Why do we not use the `children` prop instead? Is the intention for subheader supposed to not be scrollable? Do you think we have enough usecases for a subheader prop? Just want to make sure we are not getting into the habit of bloating the SidebarContent component

File: src/elements/content-sidebar/MetadataSidebarRedesign.scss [hunk 4]
```diff
@@ -4,7 +4,6 @@
     border-left: 1px solid tokens.$gray-10;
 
     .bcs-MetadataSidebarRedesign-content {
-        position: absolute;
```
Comment (JChan106) [inline src/elements/content-sidebar/MetadataSidebarRedesign.scss:7]: Why did we need to remove this?

Comment (kajarosz) [reply src/elements/content-sidebar/MetadataSidebarRedesign.scss:7]: We don't need it. I was testing sth ad forgot to bring it back. Thanks for noticing! I am getting rid of this change.

File: src/elements/content-sidebar/MetadataSidebarRedesign.tsx [hunk 5]
```diff
@@ -153,29 +154,44 @@ function MetadataSidebarRedesign({ api, elementId, fileId, onError, isFeatureEna
         }
     };
 
+    const isFullyLoaded = file && templates && templateInstances;
+    const visibleTemplateInstances = templateInstances.filter(templateInstance => !templateInstance.hidden);
+
+    const showLoading = status === STATUS.LOADING;
+    const showEmptyState = !showLoading && isFullyLoaded && visibleTemplateInstances.length === 0 && !editingTemplate;
+    const showEditor = !showEmptyState && editingTemplate;
+    const showList = !showEditor && visibleTemplateInstances.length > 0 && !editingTemplate;
```
Comment (wpiesiak) [inline src/elements/content-sidebar/MetadataSidebarRedesign.tsx:163]: Can we simplify this definitions? eg. `showList` checks `!showEditor` and `!editingTemplate` which is already checked in `showEditor`.

**Review by wpiesiak — APPROVED** [review]

File: src/elements/content-sidebar/MetadataSidebarRedesign.tsx (hunk 5 above)
Comment (kajarosz) [reply src/elements/content-sidebar/MetadataSidebarRedesign.tsx]: done!

**Review by JChan106 — APPROVED** [review]

**Review by wpiesiak — APPROVED** [review]

**Review by JChan106 — APPROVED** [review]

**Review by greg-in-a-box — APPROVED** [review]

**Pull request merged**